wire-ios-sync-engine/Source/UserSession/ZMUserSession+CoreCrypto.swift

```diff
@@ -0,0 +1,92 @@
+//
+// Wire
+// Copyright (C) 2022 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireDataModel
+import WireRequestStrategy
+import CoreCryptoSwift
+
+extension ZMUserSession {
+
+    func setupCryptoStack() {
+        syncContext.performAndWait {
+            let factory = CoreCryptoFactory()
+
+            do {
+                let configuration = try factory.createConfiguration(
+                    sharedContainerURL: sharedContainerURL,
+                    selfUser: .selfUser(in: syncContext)
+                )
+
+                let coreCrypto = try syncContext.coreCrypto ?? factory.createCoreCrypto(with: configuration)
+                syncContext.coreCrypto = coreCrypto
+
+                try createProteusServiceIfNeeded(coreCrypto: coreCrypto)
+
+                try createMLSControllerIfNeeded(
+                    coreCrypto: coreCrypto,
+                    clientID: configuration.clientID
+                )
+            } catch let error as MLSControllerSetupFailure {
+                WireLogger.coreCrypto.error("fail: setup MLSController: \(String(describing: error))")
+            } catch {
+                WireLogger.coreCrypto.error("fail: setup crypto stack: \(String(describing: error))")
+            }
+
+            WireLogger.coreCrypto.info("success: setup crypto stack")
+        }
+    }
+
+    // MARK: - Proteus
+
+    private func createProteusServiceIfNeeded(coreCrypto: CoreCryptoProtocol) throws {
+        guard syncContext.proteusService == nil else { return }
+        syncContext.proteusService = ProteusService(coreCrypto: coreCrypto)
+    }
+
+    // MARK: - MLS
+
+    private func createMLSControllerIfNeeded(
+        coreCrypto: CoreCryptoProtocol,
+        clientID: String
+    ) throws {
+        guard syncContext.mlsController == nil else { return }
+
+        guard let syncStatus = syncStatus else {
+            throw MLSControllerSetupFailure.missingSyncStatus
+        }
+
+        guard let userDefaults = UserDefaults(suiteName: "com.wire.mls.\(clientID)") else {
+            throw MLSControllerSetupFailure.invalidUserDefaults
+        }
+
+        syncContext.mlsController = MLSController(
```

Comment:
Can a NSManagedObject of a given Entity be on the uiContext and syncContext?
Just wondering when we access mlsController or the proteusService, how safe it is to retrieve it?

Reply:
Yes both contexts are used, but in different ways. The view context is primarily read only to display data in the ui. The sync context is reserved for business logic and we require that some work only be available on the sync context.

Any core crypto work should be done only on the sync context, we'll crash otherwise.